skyportal/handlers/api/internal/util.py

```diff
@@ -0,0 +1,153 @@
+import datetime
+
+import numpy as np
+from astropy import units as u
+from astropy.coordinates import SkyCoord
+from astroquery.gaia import Gaia
+from astropy.time import Time
+
+
+def get_nearby_offset_stars(source_ra, source_dec, source_name,
+                            how_many=3,
+                            radius_degrees=2 / 60.,  # 2 arcmin radius
+                            mag_limit=18.0,
+                            mag_min=10.0,
+                            min_sep_arcsec=5,
+                            starlist_type='Keck',
+                            obstime_isoformat=None,
+                            use_source_pos_in_starlist=True,
+                            remaining_searches=1
+                            ):
+
+    # TODO make these search parameters part of the API call
+    if not obstime_isoformat:
+        source_obstime = Time(datetime.datetime.utcnow().isoformat())
+    else:
+        # TODO: check the obstime format
+        source_obstime = Time(obstime_isoformat)
+    gaia_obstime = "J2015.5"
+
+    center = SkyCoord(source_ra, source_dec, unit=(u.degree, u.degree),
+                      frame='icrs', obstime=source_obstime)
+    # get three times as many stars as requested for now
+    # and go fainter as well
+    fainter_diff = 2.0  # mag
+    search_multipler = 10
+    query_string = f"""
+                  SELECT TOP {how_many*search_multipler} DISTANCE(
+                    POINT('ICRS', ra, dec),
+                    POINT('ICRS', {source_ra}, {source_dec})) AS
+                    dist, source_id, ra, dec, ref_epoch,
+                    phot_rp_mean_mag, pmra, pmdec, parallax
+                  FROM gaiadr2.gaia_source
+                  WHERE 1=CONTAINS(
+                    POINT('ICRS', ra, dec),
+                    CIRCLE('ICRS', {source_ra}, {source_dec},
+                           {radius_degrees}))
+                  AND phot_rp_mean_mag < {mag_limit + fainter_diff}
+                  AND phot_rp_mean_mag > {mag_min}
+                  AND parallax < 250
+                  ORDER BY phot_rp_mean_mag ASC
+                """
+    # TODO possibly: save the offset data (cache)
+    job = Gaia.launch_job(query_string)
+    r = job.get_results()
+
+    catalog = SkyCoord.guess_from_table(r)
+
+    # star needs to be this far away
+    # from another star
+    min_sep = min_sep_arcsec*u.arcsec
+    good_list = []
+    for source in r:
+        c = SkyCoord(ra=source["ra"], dec=source["dec"],
+                     unit=(u.degree, u.degree),
+                     pm_ra_cosdec=np.cos(source["dec"]*np.pi/180.0)*source['pmra'] * u.mas/u.yr,
+                     pm_dec=source["pmdec"] * u.mas/u.yr,
+                     frame='icrs', distance=min(abs(1/source["parallax"]),10)*u.kpc,
+                     obstime=gaia_obstime)
+
+        d2d = c.separation(catalog)  # match it to the catalog
+        if sum(d2d < min_sep) == 1 and source["phot_rp_mean_mag"] <= mag_limit:
```

Comment:
@dannygoldstein this is the protection against binaries. `min_sep` is ~5 arcsec. So if only the star matches then this is considered non-binary.